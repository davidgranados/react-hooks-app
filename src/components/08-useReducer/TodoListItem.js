import React from "react";
import classNames from "classnames";

export const TodoListItem = ({ todo, index, onDelete, onToggle }) => {
  const handleDelete = ({ target }) => {
    onDelete(parseInt(target.dataset.id));
  };
  const handleToggle = ({ target }) => {
    onToggle(parseInt(target.dataset.id));
  };
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={classNames({ complete: todo.done })}
        data-id={todo.id}
        onClick={handleToggle}
      >
        {index}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        data-id={todo.id}
        onClick={handleDelete}
      >
        Borrar
      </button>
    </li>
  );
};
