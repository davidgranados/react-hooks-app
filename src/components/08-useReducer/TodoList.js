import React from "react";
import { TodoListItem } from "./TodoListItem";

export const TodoList = ({ todoList, onDelete, onToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {todoList.map((todo, i) => (
        <TodoListItem
          key={todo.id}
          todo={todo}
          index={i + 1}
          onDelete={onDelete}
          onToggle={onToggle}
        />
      ))}
    </ul>
  );
};
