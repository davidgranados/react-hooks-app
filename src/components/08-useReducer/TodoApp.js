import React, { useReducer, useEffect } from "react";
import { todoReducer } from "./todoReducer";

import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("todoList")) || [];
};

export const TodoApp = () => {
  const [todoList, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todoList", JSON.stringify(todoList));
  }, [todoList]);

  const onDelete = (todoId) => {
    dispatch({
      type: "delete",
      payload: todoId,
    });
  };

  const onToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId,
    });
  };

  const onAdd = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo,
    });
  };

  return (
    <div>
      <h1>TodoApp ( {todoList.length} ) </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList todoList={todoList} onDelete={onDelete} onToggle={onToggle} />
        </div>

        <div className="col-5">
          <TodoAdd onAdd={onAdd} />
        </div>
      </div>
    </div>
  );
};
