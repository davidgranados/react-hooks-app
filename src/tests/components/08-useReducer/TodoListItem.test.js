import React from "react";
import { shallow } from "enzyme";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodo } from "../../fixtures/demoTodo";

//demoTodo

describe("Pruebas en <TodoListItem />", () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const wrapper = shallow(
    <TodoListItem
      todo={demoTodo[0]}
      index={0}
      onDelete={onDelete}
      onToggle={onToggle}
    />
  );

  it("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("debe de llamar la función handleDelete", () => {
    const button = wrapper.find("button");
    const clickEvent = {
      target: {
        dataset: {
          id: button.prop("data-id"),
        },
      },
    };
    button.simulate("click", clickEvent);
    expect(onDelete).toHaveBeenCalledTimes(1);
    expect(onDelete).toHaveBeenCalledWith(
      parseInt(clickEvent.target.dataset.id)
    );
  });

  it("debe de llamar la función onToggle", () => {
    const p = wrapper.find("p");
    const clickEvent = {
      target: {
        dataset: {
          id: p.prop("data-id"),
        },
      },
    };
    p.simulate("click", clickEvent);
    expect(onToggle).toHaveBeenCalledWith(
      parseInt(clickEvent.target.dataset.id)
    );
  });

  it("debe de mostrar el texto correctamente", () => {
    const p = wrapper.find("p");
    expect(p.text().trim()).toBe(`0. ${demoTodo[0].desc}`);
  });

  it("debe de tener la clase complete si el TODO.done = true", () => {
    const todo = demoTodo[0];
    todo.done = true;
    const wrapper = shallow(<TodoListItem todo={todo} />);
    expect(wrapper.find("p").hasClass("complete")).toBe(true);
  });
});
