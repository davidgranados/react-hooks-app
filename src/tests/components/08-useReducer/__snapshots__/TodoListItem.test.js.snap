// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem /> debe de mostrarse correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className=""
    data-id={1}
    onClick={[Function]}
  >
    0
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    data-id={1}
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
